model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProvider.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.models.map.storage.jpa;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import javax.persistence.EntityManager;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.map.common.AbstractEntity;
+import org.keycloak.models.map.storage.MapKeycloakTransaction;
+import org.keycloak.models.map.storage.MapStorage;
+import org.keycloak.models.map.storage.MapStorageProvider;
+import org.keycloak.models.map.storage.MapStorageProviderFactory.Flag;
+import org.keycloak.models.map.storage.jpa.client.JpaClientMapKeycloakTransaction;
+
+public class JpaMapStorageProvider implements MapStorageProvider {
+
+    private final JpaMapStorageProviderFactory factory;
+    private final KeycloakSession session;
+    private final EntityManager em;
+    private final Map<Class<?>, Function<EntityManager, MapKeycloakTransaction>> transactionConstructors = new HashMap<>();
+
+    public JpaMapStorageProvider(JpaMapStorageProviderFactory factory, KeycloakSession session, EntityManager em) {
+        this.factory = factory;
+        this.session = session;
+        this.em = em;
+        transactionConstructors.put(ClientModel.class, JpaClientMapKeycloakTransaction::new);
+    }
+
+    @Override
+    public void close() {
+        em.close();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <V extends AbstractEntity, M> MapStorage<V, M> getStorage(Class<M> modelType, Flag... flags) {
+        factory.validateAndUpdateSchema(session, modelType);
+        return (session) -> transactionConstructors.get(modelType).apply(em);
```

Comment:
- There is no guarantee that `MapStorage` would remain a single-method interface in the future, and likely there will be further methods added. Please use a proper (inner or just anonymous) class
- I realized we have a bug in logic in this method, #9321. Please keep track of created transactions in the storage per model in the `session` and just return an existing one if called multiple times. See e.g. https://github.com/keycloak/keycloak/blob/96b2669a00dc1396138f863bbc11aa0b0aa3f0b6/model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapStorage.java#L131

Reply:
I've tested it a little bit and it seems that currently for clients it's not called multiple times within the same session. I assume it could change in future or other stores could call it multiple times, right?

Reply:
Exactly